Grouping Statement: Cannot bind argument at index N+1 because the index is out of range.  The statement has N parameters.

**Runtime Environment**
SQLDelight version: 1.4.0
Application OS: Android

**Describe the bug**
I'm trying to convert a separate update + insert queries into a single upsert query using the Grouping Statement feature.

Consider the following schema:
```sql
CREATE TABLE Bug (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    col0 TEXT NOT NULL,
    col1 TEXT NOT NULL,
    col2 INTEGER AS Boolean NOT NULL DEFAULT 0,
    col3 INTEGER NOT NULL DEFAULT (strftime('%s', 'now'))
);

CREATE UNIQUE INDEX Bug__col0_col1 ON Bug(col0, col1);

upsert {
  UPDATE Bug
  SET col2 = :col2,
      col3 = :col3
  WHERE col0 = :col0
    AND col1 = :col1
  ;

  INSERT OR IGNORE INTO Bug(col0, col1, col2)
  VALUES (:col0, :col1, :col2)
  ;
}
```
When running the following query like so
```kotlin
        val driver = AndroidSqliteDriver(MyCoolStore.Schema, this, null)
        val db = MyCoolStore(driver)
        db.bugQueries.upsert(
            col0 = "a",
            col1 = "b",
            col2 = true,
            col3 = 123
        )
```
SqlDelight throws up:
```
     Caused by: java.lang.IllegalArgumentException: Cannot bind argument at index 5 because the index is out of range.  The statement has 4 parameters.
        at android.database.sqlite.SQLiteProgram.bind(SQLiteProgram.java:212)
        at android.database.sqlite.SQLiteProgram.bindString(SQLiteProgram.java:166)
        at androidx.sqlite.db.framework.FrameworkSQLiteProgram.bindString(FrameworkSQLiteProgram.java:50)
        at com.squareup.sqldelight.android.AndroidPreparedStatement.bindString(AndroidSqliteDriver.kt:189)
        at com.littledot.androidbugreport.app.BugQueriesImpl$upsert$1.invoke(MyCoolStoreImpl.kt:118)
        at com.littledot.androidbugreport.app.BugQueriesImpl$upsert$1.invoke(MyCoolStoreImpl.kt:92)
        at com.squareup.sqldelight.android.AndroidSqliteDriver.execute(AndroidSqliteDriver.kt:120)
        at com.squareup.sqldelight.android.AndroidSqliteDriver.execute(AndroidSqliteDriver.kt:136)
        at com.littledot.androidbugreport.app.BugQueriesImpl.upsert(MyCoolStoreImpl.kt:102)
        at com.littledot.androidbugreport.MainActivity.onCreate(MainActivity.kt:17)
```
BugQueriesImpl.upsert:
```kotlin
private class BugQueriesImpl(
  private val database: MyCoolStoreImpl,
  private val driver: SqlDriver
) : TransacterImpl(driver), BugQueries {
  override fun upsert(
    col2: Boolean,
    col3: Long,
    col0: String,
    col1: String
  ) {
    driver.execute(852179207, """
    |UPDATE Bug
    |  SET col2 = ?,
    |      col3 = ?
    |  WHERE col0 = ?
    |    AND col1 = ?
    |  ;
    |
    |  INSERT OR IGNORE INTO Bug(col0, col1, col2)
    |  VALUES (?, ?, ?)
    |  ;
    """.trimMargin(), 7) {
      bindLong(1, if (col2) 1L else 0L)
      bindLong(2, col3)
      bindString(3, col0)
      bindString(4, col1)
      bindString(5, col0)
      bindString(6, col1)
      bindLong(7, if (col2) 1L else 0L)
    }
  }
}
```
Repro: Run project https://gitlab.com/littledot/android-bug-report

--- a/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/compiler/QueryGenerator.kt
+++ b/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/compiler/QueryGenerator.kt
@@ -2,20 +2,25 @@ package com.squareup.sqldelight.core.compiler
 
 import com.alecstrong.sql.psi.core.psi.SqlBinaryEqualityExpr
 import com.alecstrong.sql.psi.core.psi.SqlBindExpr
+import com.alecstrong.sql.psi.core.psi.SqlStmt
 import com.alecstrong.sql.psi.core.psi.SqlTypes
 import com.intellij.psi.PsiElement
 import com.intellij.psi.PsiWhiteSpace
+import com.intellij.psi.util.PsiTreeUtil
 import com.squareup.kotlinpoet.CodeBlock
 import com.squareup.kotlinpoet.FunSpec
 import com.squareup.kotlinpoet.NameAllocator
 import com.squareup.sqldelight.core.compiler.integration.javadocText
 import com.squareup.sqldelight.core.compiler.model.BindableQuery
+import com.squareup.sqldelight.core.compiler.model.NamedExecute
 import com.squareup.sqldelight.core.compiler.model.NamedQuery
 import com.squareup.sqldelight.core.lang.DRIVER_NAME
 import com.squareup.sqldelight.core.lang.util.childOfType
+import com.squareup.sqldelight.core.lang.util.findChildrenOfType
 import com.squareup.sqldelight.core.lang.util.isArrayParameter
 import com.squareup.sqldelight.core.lang.util.range
 import com.squareup.sqldelight.core.lang.util.rawSqlText
+import com.squareup.sqldelight.core.psi.SqlDelightStmtClojureStmtList
 
 abstract class QueryGenerator(private val query: BindableQuery) {
   /**
@@ -36,35 +41,56 @@ abstract class QueryGenerator(private val query: BindableQuery) {
   protected fun executeBlock(): CodeBlock {
     val result = CodeBlock.builder()
 
-    val precedingArrays = mutableListOf<String>()
+    if (query is NamedExecute && query.statement is SqlDelightStmtClojureStmtList) {
+      query.statement.findChildrenOfType<SqlStmt>().forEachIndexed { index, statement ->
+        result.add(executeBlock(statement, query.idForIndex(index)))
+      }
+    } else {
+      result.add(executeBlock(query.statement, query.id))
+    }
+
+    return result.build()
+  }
+
+  private fun executeBlock(
+    statement: PsiElement,
+    id: Int
+  ): CodeBlock {
+    val result = CodeBlock.builder()
+
+    val positionToArgument = mutableListOf<Triple<Int, BindableQuery.Argument, SqlBindExpr?>>()
+    query.arguments.forEach { argument ->
+      if (argument.bindArgs.isNotEmpty()) {
+        argument.bindArgs
+            .filter { PsiTreeUtil.isAncestor(statement, it, true) }
+            .forEach { bindArg ->
+              positionToArgument.add(Triple(bindArg.node.textRange.startOffset, argument, bindArg))
+            }
+      } else {
+        positionToArgument.add(Triple(0, argument, null))
+      }
+    }
+
     val bindStatements = CodeBlock.builder()
     val replacements = mutableListOf<Pair<IntRange, String>>()
     val argumentCounts = mutableListOf<String>()
 
     var needsFreshStatement = false
 
-    val positionToArgument = mutableListOf<Triple<Int, BindableQuery.Argument, SqlBindExpr?>>()
-    query.arguments.forEach { argument ->
-      if (argument.bindArgs.isNotEmpty()) {
-        argument.bindArgs.forEach { bindArg ->
-          positionToArgument.add(Triple(bindArg.node.textRange.startOffset, argument, bindArg))
-        }
-      } else {
-        positionToArgument.add(Triple(0, argument, null))
-      }
-    }
-
     val seenArrayArguments = mutableSetOf<BindableQuery.Argument>()
-
-    // A list of [SqlBindExpr] in order of appearance in the query.
-    val orderedBindArgs = positionToArgument.sortedBy { it.first }
-
-    // The number of non-array bindArg's we've encountered so far
-    var nonArrayBindArgsCount = 0
 
     val argumentNameAllocator = NameAllocator().apply {
       query.arguments.forEach { newName(it.type.name) }
     }
+
+    // A list of [SqlBindExpr] in order of appearance in the query.
+    val orderedBindArgs = positionToArgument.sortedBy { it.first }
+
+    // The number of non-array bindArg's we've encountered so far.
+    var nonArrayBindArgsCount = 0
+
+    // A list of arrays we've encountered so far.
+    val precedingArrays = mutableListOf<String>()
 
     // For each argument in the sql
     orderedBindArgs.forEach { (_, argument, bindArg) ->
@@ -133,8 +159,6 @@ abstract class QueryGenerator(private val query: BindableQuery) {
       }
     }
 
-    val id = if (needsFreshStatement) "null" else "${query.id}"
-
     // Adds the actual SqlPreparedStatement:
     // statement = database.prepareStatement("SELECT * FROM test")
     val executeMethod = if (query is NamedQuery) {
@@ -146,7 +170,7 @@ abstract class QueryGenerator(private val query: BindableQuery) {
       argumentCounts.add(0, nonArrayBindArgsCount.toString())
     }
     val arguments = mutableListOf<Any>(
-        query.statement.rawSqlText(replacements),
+        statement.rawSqlText(replacements),
         argumentCounts.ifEmpty { listOf(0) }.joinToString(" + ")
     )
     val binder: String
@@ -163,7 +187,12 @@ abstract class QueryGenerator(private val query: BindableQuery) {
           .build())
       binder = "%L"
     }
-    result.add("$executeMethod($id, %P, %L)$binder\n", *arguments.toTypedArray())
+    result.add("$executeMethod(" +
+        "${if (needsFreshStatement) "null" else "$id"}," +
+        " %P," +
+        " %L" +
+        ")$binder\n", *arguments.toTypedArray()
+    )
 
     return result.build()
   }

--- a/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/compiler/model/NamedExecute.kt
+++ b/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/compiler/model/NamedExecute.kt
@@ -14,6 +14,12 @@ open class NamedExecute(
     // the sqlFile package name -> com.example.
     // sqlFile.name -> test.sq
     // name -> query name
-    get() = getUniqueQueryIdentifier(
-        statement.sqFile().let { "${it.packageName}:${it.name}:$name" })
+    get() = idForIndex(null)
+
+  internal fun idForIndex(index: Int?): Int {
+    val postFix = if (index == null) "" else "_$index"
+    return getUniqueQueryIdentifier(statement.sqFile().let {
+      "${it.packageName}:${it.name}:$name$postFix"
+    })
+  }
 }

--- a/sqldelight-compiler/src/test/kotlin/com/squareup/sqldelight/core/queries/SelectQueryTypeTest.kt
+++ b/sqldelight-compiler/src/test/kotlin/com/squareup/sqldelight/core/queries/SelectQueryTypeTest.kt
@@ -690,16 +690,17 @@ class SelectQueryTypeTest {
 
     assertThat(generator.function().toString()).isEqualTo("""
       |override fun insertTwice(value: kotlin.Long) {
-      |  driver.execute(${query.id}, ""${'"'}
+      |  driver.execute(${query.idForIndex(0)}, ""${'"'}
       |  |INSERT INTO data (value)
       |  |  VALUES (?)
-      |  |  ;
-      |  |  INSERT INTO data (value)
-      |  |  VALUES (?)
-      |  |  ;
-      |  ""${'"'}.trimMargin(), 2) {
+      |  ""${'"'}.trimMargin(), 1) {
       |    bindLong(1, value)
-      |    bindLong(2, value)
+      |  }
+      |  driver.execute(${query.idForIndex(1)}, ""${'"'}
+      |  |INSERT INTO data (value)
+      |  |  VALUES (?)
+      |  ""${'"'}.trimMargin(), 1) {
+      |    bindLong(1, value)
       |  }
       |}
       |""".trimMargin())
@@ -728,16 +729,17 @@ class SelectQueryTypeTest {
 
     assertThat(generator.function().toString()).isEqualTo("""
       |override fun insertTwice(value: kotlin.Long, value_: kotlin.Long) {
-      |  driver.execute(${query.id}, ""${'"'}
+      |  driver.execute(${query.idForIndex(0)}, ""${'"'}
       |  |INSERT INTO data (value)
       |  |  VALUES (?)
-      |  |  ;
-      |  |  INSERT INTO data (value)
-      |  |  VALUES (?)
-      |  |  ;
-      |  ""${'"'}.trimMargin(), 2) {
+      |  ""${'"'}.trimMargin(), 1) {
       |    bindLong(1, value)
-      |    bindLong(2, value_)
+      |  }
+      |  driver.execute(${query.idForIndex(1)}, ""${'"'}
+      |  |INSERT INTO data (value)
+      |  |  VALUES (?)
+      |  ""${'"'}.trimMargin(), 1) {
+      |    bindLong(1, value_)
       |  }
       |}
       |""".trimMargin())
@@ -766,16 +768,17 @@ class SelectQueryTypeTest {
 
     assertThat(generator.function().toString()).isEqualTo("""
       |override fun insertTwice(value: kotlin.Long) {
-      |  driver.execute(${query.id}, ""${'"'}
+      |  driver.execute(${query.idForIndex(0)}, ""${'"'}
       |  |INSERT INTO data (value)
       |  |  VALUES (?)
-      |  |  ;
-      |  |  INSERT INTO data (value)
-      |  |  VALUES (?)
-      |  |  ;
-      |  ""${'"'}.trimMargin(), 2) {
+      |  ""${'"'}.trimMargin(), 1) {
       |    bindLong(1, value)
-      |    bindLong(2, value)
+      |  }
+      |  driver.execute(${query.idForIndex(1)}, ""${'"'}
+      |  |INSERT INTO data (value)
+      |  |  VALUES (?)
+      |  ""${'"'}.trimMargin(), 1) {
+      |    bindLong(1, value)
       |  }
       |}
       |""".trimMargin())
@@ -804,16 +807,17 @@ class SelectQueryTypeTest {
 
     assertThat(generator.function().toString()).isEqualTo("""
       |override fun insertTwice(value: kotlin.Long, value_: kotlin.Long) {
-      |  driver.execute(${query.id}, ""${'"'}
+      |  driver.execute(${query.idForIndex(0)}, ""${'"'}
       |  |INSERT INTO data (value)
       |  |  VALUES (?)
-      |  |  ;
-      |  |  INSERT INTO data (value)
-      |  |  VALUES (?)
-      |  |  ;
-      |  ""${'"'}.trimMargin(), 2) {
+      |  ""${'"'}.trimMargin(), 1) {
       |    bindLong(1, value)
-      |    bindLong(2, value_)
+      |  }
+      |  driver.execute(${query.idForIndex(1)}, ""${'"'}
+      |  |INSERT INTO data (value)
+      |  |  VALUES (?)
+      |  ""${'"'}.trimMargin(), 1) {
+      |    bindLong(1, value_)
       |  }
       |}
       |""".trimMargin())
@@ -846,16 +850,17 @@ class SelectQueryTypeTest {
 
     assertThat(generator.function().toString()).isEqualTo("""
       |override fun insertTwice(value: kotlin.Long, value_: kotlin.Long) {
-      |  driver.execute(${query.id}, ""${'"'}
+      |  driver.execute(${query.idForIndex(0)}, ""${'"'}
       |  |INSERT INTO data (value)
       |  |  VALUES (?)
-      |  |  ;
-      |  |  INSERT INTO data (value)
-      |  |  VALUES (?)
-      |  |  ;
-      |  ""${'"'}.trimMargin(), 2) {
+      |  ""${'"'}.trimMargin(), 1) {
       |    bindLong(1, value)
-      |    bindLong(2, value_)
+      |  }
+      |  driver.execute(${query.idForIndex(1)}, ""${'"'}
+      |  |INSERT INTO data (value)
+      |  |  VALUES (?)
+      |  ""${'"'}.trimMargin(), 1) {
+      |    bindLong(1, value_)
       |  }
       |  notifyQueries(${query.id}, {database.testQueries.someSelect})
       |}

--- a/sqldelight-gradle-plugin/src/test/integration/src/main/sqldelight/com/squareup/sqldelight/integration/GroupedStatement.sq
+++ b/sqldelight-gradle-plugin/src/test/integration/src/main/sqldelight/com/squareup/sqldelight/integration/GroupedStatement.sq
@@ -1,0 +1,25 @@
+CREATE TABLE Bug (
+    col0 TEXT NOT NULL,
+    col1 TEXT NOT NULL,
+    col2 INTEGER AS Boolean NOT NULL DEFAULT 0,
+    col3 INTEGER NOT NULL
+);
+
+CREATE UNIQUE INDEX Bug__col0_col1 ON Bug(col0, col1);
+
+upsert {
+  UPDATE Bug
+  SET col2 = :col2,
+      col3 = :col3
+  WHERE col0 = :col0
+    AND col1 = :col1
+  ;
+
+  INSERT OR IGNORE INTO Bug(col0, col1, col2, col3)
+  VALUES (:col0, :col1, :col2, :col3)
+  ;
+}
+
+selectAll:
+SELECT *
+FROM Bug;

--- a/sqldelight-gradle-plugin/src/test/integration/src/test/java/com/squareup/sqldelight/integration/IntegrationTests.kt
+++ b/sqldelight-gradle-plugin/src/test/integration/src/test/java/com/squareup/sqldelight/integration/IntegrationTests.kt
@@ -18,6 +18,7 @@ class IntegrationTests {
   private lateinit var nullableTypesQueries: NullableTypesQueries
   private lateinit var bigTableQueries: BigTableQueries
   private lateinit var varargsQueries: VarargsQueries
+  private lateinit var groupedStatementQueries: GroupedStatementQueries
 
   private object listAdapter : ColumnAdapter<List<String>, String> {
     override fun decode(databaseValue: String): List<String> = databaseValue.split(",")
@@ -34,6 +35,7 @@ class IntegrationTests {
     nullableTypesQueries = queryWrapper.nullableTypesQueries
     bigTableQueries = queryWrapper.bigTableQueries
     varargsQueries = queryWrapper.varargsQueries
+    groupedStatementQueries = queryWrapper.groupedStatementQueries
   }
 
   @Test fun indexedArgs() {
@@ -143,6 +145,17 @@ class IntegrationTests {
     assertThat(varargsQueries.select(setOf(1, 2), 10).executeAsList()).containsExactly(
         MyTable(1, 1, 10),
         MyTable(2, 2, 10)
+    )
+  }
+
+  @Test fun groupedStatement() {
+    groupedStatementQueries.upsert(col0 = "1", col1 = "1", col2 = true, col3 = 10)
+    groupedStatementQueries.upsert(col0 = "2", col1 = "2", col2 = true, col3 = 20)
+    groupedStatementQueries.upsert(col0 = "1", col1 = "1", col2 = false, col3 = 11)
+
+    assertThat(groupedStatementQueries.selectAll().executeAsList()).containsExactly(
+        Bug("2", "2", true, 20),
+        Bug("1", "1", false, 11)
     )
   }
 }

